timesketch/lib/analyzers/hashr_lookup.py

```diff
@@ -0,0 +1,317 @@
+          sources: A list of sources (image names) where this hash is known from.
+          event:  The OpenSearch event object that contains this hash and needs
+                  to be tagged or to add an attribute.
+
+        Returns:
+          None: If everything worked the function just returns.
+        """
+        tags_container = ['known-hash']
+        if (hash_value ==
+            'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
+            ):
+            tags_container.append('zerobyte-file')
+            self.zerobyte_file_counter += 1
+            # Do not add any source attribute for zerobyte files,
+            # since it exists in all sources.
+            sources = False
+
+        event.add_tags(tags_container)
+
+        if sources:
+            event.add_attributes({'hashR_sample_sources': list(sources)})
+
+        event.commit()
+
+    def run(self):
+        """Entry point for the analyzer.
+
+        Returns:
+            String with summary of the analyzer result
+        """
+        # Check if analyzer can connect to hashR:
+        msg = self.connect_hashR()
+        if isinstance(msg, str):
+            return msg
+
+        # Note: Add fieldnames that contain sha256 values in your events.
+        query = ('_exists_:hash_sha256 OR _exists_:sha256 OR _exists_:hash OR '
+                 '_exists_:sha256_hash')
+
+        # Note: Add fieldnames that contain sha256 values in your events.
+        return_fields = ['hash_sha256', 'hash', 'sha256', 'sha256_hash']
+
+        # Generator of events based on your query.
+        # Swap for self.event_pandas to get pandas back instead of events.
+        events = self.event_stream(
+            query_string=query, return_fields=return_fields)
+        known_hash_counter = 0
+        error_hash_counter = 0
+        total_event_counter = 0
+        hash_events_dict = {}
+        logger.info('Collect a list of unique hashes to check against hashR.')
+        for event in events:
+            total_event_counter += 1
+            hash_value = None
+            for key in return_fields:
+                if key in event.source.keys():
+                    hash_value = event.source.get(key)
+                    break
+            if not hash_value:
+                error_hash_counter += 1
+                logger.warning(
+                    'No matching field with a hash found in this event! '
+                    '-- Event Source: %s', str(event.source))
+                continue
+
+            if len(hash_value) != 64:
+                logger.warning('The extracted hash does not match the required '
+                               'lenght (64) of a SHA256 hash. Skipping this '
+                               'event! Hash: %s - Lenght: %d',
+                               hash_value, len(hash_value))
+                error_hash_counter += 1
+                continue
+
+            hash_events_dict.setdefault(hash_value, []).append(event)
```

Comment:
We are already in a generator from the event_stream. Instead of building up a big dict in memory we can chunk it up here and check against hashR. The check_against_hashr() can be refactored to yield (a generator) events that matches. Then you annotate those and let the datastore abstraction flush events as necessary.

```
    CHUNK_SIZE = 50000
    events_with_hash_value = []

    for event in events:
            event_counter += 1
            hash_value = None

            for key in return_fields:
                if key in event.source.keys():
                    hash_value = event.source.get(key)
                    break

            if not hash_value:
                continue

            if len(hash_value) != 64:
                continue

            events_with_hash_value.append(event)

            if event_counter % int(CHUNK_SIZE) == 0:
                for matching_event in check_against_hashr(events_with_hash_value):  <-- make check_against_hashr() return a generator of events that have known hashes
                    annotate_event(sample_hash, False, event)
                    events_with_hash_value = []
                    events_counter = 0

         
       # Import the remaining events.
       if len(events_with_hash_value) > 0:
           for matching_event in check_against_hashr(events_with_hash_value):
               annotate_event(sample_hash, False, event)
```

Reply:
As discussed offline: Since this generator approach does not work well with the deduplication of hashes and the current principal of only query each hash once against the hashR DB, we will postpone this for to a future version of the analyzer if necessary. If the current solution shows to be problematic in real-life investigations we can use this as an alternative approach.